Edits to CHM to remove compiler warnings on Apple Clang 17.0.0 on Intel MacOSX Sequoia 15.7.2

djmallum opened the pull request:
I was bored and decided to deal with the warnings that CHM produces during compilation. On my machine, CHM no longer produces compiler warns in the develop branch.

All of these should not impact the code in any way, however, the largest change is in src/utility/str_format.h and should be reviewed/tested to ensure that it encodes the same behaviour.

Finally, I went with removing variables that were set but not used, rather than commenting them out. Can easily revert to just commented out if desired.

djmallum committed:
Third Party: Edited snowpack functions to avoid warnings

Variables set but unused, deleted lines.

djmallum committed:
Removed set but unused variable chunks

djmallum committed:
Fixed Boost warning

`boost::make_shared_object<T>` prints a warning if if `T` has virtual
functions it must have a virtual deconstructor.

djmallum committed:
Fixed order of initialization

djmallum committed:
Removed variable that is set but unused

djmallum committed:
Removed variable that is set but unused

djmallum committed:
Fixed variable length array warning

VLAs are a GCC and Clang extension. Rewrote the function following the
linked comment. Removed link, linked to the same stackoverflow question.
The new version is simply c++11 complaint. The C++20 version is not
supported in all compilers yet.

djmallum commented:
Dealt with the two comments from copilot. Ready to go.

Chrismarsh commented on src/utility/str_format.h:
```diff
@@ -28,20 +28,25 @@
 
 
 
+#include <stdexcept>
```
Can you just delete this file and replace it with std::format?
https://stackoverflow.com/a/26221725

the only use of this function is _determine_modeul_dep L2074

djmallum replied:
You mean delete the function and file and just use it in the one place it is used?

Chrismarsh replied:
I think str_format is only used here
https://github.com/Chrismarsh/CHM/blob/92ade51f6ac01442b88a9b94dea81e3fbeb6f5be/src/core.cpp#L2209

Can you remove str_format.h entirely and replace with std::format like described here? https://stackoverflow.com/questions/2342162/stdstring-formatting-like-sprintf/26221725#26221725

Chrismarsh commented on src/modules/snow_slide.cpp:
Delete here too

Chrismarsh replied:
Replying to note this in the reivew

Chrismarsh commented on third_party/snowpack/Utils.cc:
```diff
@@ -451,14 +435,14 @@ bool massBalanceCheck(const SnowStation& Xdata, const SurfaceFluxes& Sdata, doub
 double forcedErosion(const double hs, SnowStation& Xdata)
 {
-	int    nErode=0;        // Counters
+	// int    nErode=0;        // Counters
```
Given you are removing the dead code elsewhere, either leave the other stuff commented, or remove this. I think leavning it commented is better to, in theory, make integration of next snowpack model version easier

djmallum replied:
OK I can change these changes to be commented and include a comment explaining (or maybe a git blame is enough for anything to check)

Chrismarsh replied:
Just leaving them commented out is fine imo. Snowpack is due to a new version to be integrated soon anyway

djmallum committed:
Commented out removed lines

Lines commented in this commit were triggering a set-but-not-used
compiler warning. Previous commits in this branch removed the lines
entirely. In this commit this was partially reverted so that lines
remain commented with a comment explaining in case of future use.

djmallum committed:
Replaced c-style string formatting with std::format

djmallum commented:
@Chrismarsh Can you resolve these conflicts? Looks to be related to changes you've made since I made this commit. I handled the conflict in `src/core.cpp`.

Chrismarsh commented:
Sorry about that. Can you check if this still resolves all the warnings?

djmallum committed:
Fixing unused warning in gis utility function

GDAL CreateField  member function to the OGRLayer class (a) does not
return void and (b) forces a warning if the output is ignored.

The output is now captured and thrown if it fails. Throws on all error
types and also outputs the error type. ogr_core.h needs to be checked to
see what the error type (enum) actually means.

djmallum committed:
Fixed new warnings after rebase with upstream

New warnings were result of some changes that occured before the PR has
been merged. #156

Chrismarsh committed:
Merge branch 'develop' into fix/compile_warnings_clang

Chrismarsh commented:
I resolved the merge for ugrid witer's ctor, but it might need to be tweaked?

djmallum commented on src/mesh/ugrid_writer.cpp:
```diff
@@ -16,7 +16,8 @@
 
-ugrid_writer::ugrid_writer(mesh m, boost::shared_ptr<global> g, bool write_parameters, std::string fname, bool use_zarr):
+ugrid_writer::ugrid_writer(mesh m, std::shared_ptr<global> g, bool write_parameters, std::string fname, bool use_zarr):
+    _mesh(m),
```
Almost, extra `_mesh` instantiation here. Remove this and this is ok.

Chrismarsh committed:
Remove redundant initialization of _mesh

djmallum committed:
correcting static_pointer_cast to standard library

Chrismarsh commented:
lgtm thanks
